tests/test_workers/test_tasks/test_general.py

```diff
@@ -259,25 +263,49 @@ def test_finalize_request(
     mock_create_archive,
     mock_get_request,
     task_passes_state_check,
+    retries_until_file_is_ready,
+    raise_error,
 ):
     pkg = {"name": "foo", "version": "1.0", "type": "pip"}
-    mock_get_request.return_value = {
-        "flags": ["some-flag"],
-        "pkg_managers": ["pip"],
-        "packages": [pkg],
-        "dependencies": [pkg, pkg],
-    }
 
+    # the mock will be called at least twice, once for the first loading of the request and
+    # once for the first check on the packages file
+    expected_mock_calls = retries_until_file_is_ready + 2
+
+    def side_effect(*args):
+        call_count = mock_get_request.call_count
+
+        # the mock should return a full answer only when we reach the expected amount of retries
+        if call_count == expected_mock_calls:
+            return {
+                "flags": ["some-flag"],
+                "pkg_managers": ["pip"],
+                "packages": [pkg],
+                "dependencies": [pkg, pkg],
+            }
+        else:
+            return {
+                "flags": ["some-flag"],
+                "pkg_managers": ["pip"],
+                "packages": [],
+                "dependencies": [],
+            }
+
+    mock_get_request.side_effect = side_effect
     mock_aggregate_data.return_value = mock.Mock(packages=[pkg], all_dependencies=[pkg, pkg])
 
-    tasks.finalize_request(42)
+    with raise_error and pytest.raises(CachitoError) or nullcontext():
+        tasks.finalize_request(42)
 
-    mock_get_request.assert_called_once_with(42)
+    mock_get_request.assert_called_with(42)
     mock_create_archive.assert_called_once_with(42, ["some-flag"])
     mock_save_bundle_archive_checksum.assert_called_once_with(42)
     mock_aggregate_data.assert_called_once_with(42, ["pip"])
     mock_set_counts.assert_called_once_with(42, 1, 2)
-    mock_set_state.assert_called_once_with(42, "complete", "Completed successfully")
+
+    if not raise_error:
+        assert mock_get_request.call_count == expected_mock_calls
```

Comment:
Shouldn't this assertion be true even in the error case?

Reply:
Since the test config is limited to 3 retries, it will never reach the expected retries of 5. The ideia here is just making sure the code raises an exception once the limit is over.

Reply:
oh, right